code/core/src/phone/java/com/adobe/marketing/mobile/MobileCoreInitializer.kt

```diff
@@ -0,0 +1,215 @@
+            try {
+                Date().toString()
+            } catch (e: AssertionError) {
+                // Ignore AssertionError
+            } catch (e: Exception) {
+                // Ignore other exceptions
+            }
+        }
+    }
+
+    fun setApplication(application: Application) {
+        if (!isUserUnlocked(application)) {
+            Log.error(
+                CoreConstants.LOG_TAG,
+                LOG_TAG,
+                "setApplication failed - device is in direct boot mode, SDK will not be" +
+                    " initialized."
+            )
+            return
+        }
+        handleAndroid8BugWorkaround()
+
+        if (setApplicationCalled.getAndSet(true)) {
+            Log.debug(
+                CoreConstants.LOG_TAG,
+                LOG_TAG,
+                "setApplication failed - ignoring as setApplication was already called."
+            )
+            return
+        }
+
+        ServiceProvider.getInstance().appContextService.setApplication(application)
+        App.registerActivityLifecycleCallbacks(LaunchInfoCollector())
+        scope.launch {
+            mutex.withLock {
+                try {
+                    val v4Migrator = V4Migrator()
+                    v4Migrator.migrate()
+                } catch (e: Exception) {
+                    Log.error(CoreConstants.LOG_TAG, LOG_TAG, "Migration from V4 SDK failed with error - ${e.localizedMessage}")
+                }
+
+                EventHub.shared.initializeEventHistory()
+            }
+        }
+    }
+
+    fun registerExtensions(extensions: List<Class<out Extension>?>?, completionCallback: AdobeCallback<*>?) {
+        if (!setApplicationCalled.get()) {
+            Log.error(
+                CoreConstants.LOG_TAG,
+                LOG_TAG,
+                "Failed to registerExtensions - setApplication not called"
+            )
+            return
+        }
+
+        val extensionsToRegister = mutableSetOf<Class<out Extension>>(ConfigurationExtension::class.java)
+        extensions?.filterNotNull()?.let {
+            extensionsToRegister.addAll(it)
+        }
+
+        scope.launch {
+            mutex.withLock {
+                EventHub.shared.registerExtensions(extensionsToRegister) {
+                    completionCallback?.call(null)
+                }
+            }
+        }
+    }
+
+    @VisibleForTesting
+    fun reset() {
+        setApplicationCalled.set(false)
+        initializeCalled.set(false)
+    }
+}
+
+/**
+ * Responsible for tracking the application lifecycle and automatically invoking lifecycle APIs.
+ * The lifecycle extension has internal logic to ignore lifecycle calls made during activity switching.
+ * This class handles activity counting for multi-resume scenarios and relies on the lifecycle extension
+ * to drop repeated calls as necessary.
+ */
+internal class LifecycleTracker(val additionalContextData: Map<String, String>?) : App.ActivityLifecycleCallbacks {
```

Comment:
Can we add a small note for reference that the callbacks here are invoked on the main thread and hence why they are serial.